src/index.js

```diff
@@ -150,55 +155,100 @@ class Bitswap {
+        return callback(err)
       }
-    }
 
-    const addListener = () => {
-      unwantListeners[cidStr] = () => {
-        log(`manual unwant: ${cidStr}`)
-        cleanupListener()
-        this.wm.cancelWants([cid])
+      if (blocks && blocks.length > 0) {
+        callback(null, blocks[0])
+      } else {
+        // when a unwant happens
         callback()
       }
+    })
+  }
 
-      blockListeners[cidStr] = (block) => {
-        this.wm.cancelWants([cid])
-        cleanupListener(cid)
-        callback(null, block)
-      }
+  /**
+   * Fetch a a list of blocks by cid. If the blocks are in the local
+   * blockstore they are returned, otherwise the blocks are added to the wantlist and returned once another node sends them to us.
+   *
+   * @param {Array<CID>} cids
+   * @param {function(Error, Blocks)} callback
+   * @returns {void}
+   */
+  getMany (cids, callback) {
+    const retrieved = []
+    const locals = []
+    const missing = []
+    const canceled = []
+
+    const finish = once(() => {
+      map(locals, (cid, cb) => {
+        this.blockstore.get(cid, cb)
+      }, (err, localBlocks) => {
+        if (err) {
+          return callback(err)
+        }
+
+        callback(null, localBlocks.concat(retrieved))
+      })
+    })
 
-      this.notifications.once(unwantEvent, unwantListeners[cidStr])
-      this.notifications.once(blockEvent, blockListeners[cidStr])
+    this._log('getMany', cids.length)
+
+    const addListeners = (cids) => {
+      cids.forEach((cid) => {
+        this.notifications.wantBlock(
+          cid,
+          // called on block receive
+          (block) => {
+            this.wm.cancelWants([cid])
+            retrieved.push(block)
+
+            if (retrieved.length === missing.length) {
+              finish()
+            }
+          },
+          // called on unwant
+          () => {
+            this.wm.cancelWants([cid])
+            canceled.push(cid)
+            if (canceled.length + retrieved.length === missing.length) {
+              finish()
+            }
+          }
+        )
+      })
     }
 
-    this.blockstore.has(cid, (err, has) => {
-      if (err) {
-        return callback(err)
-      }
+    each(cids, (cid, cb) => {
+      this.blockstore.has(cid, (err, has) => {
+        if (err) {
+          return cb(err)
+        }
 
-      if (has) {
-        log('already have block: %s', cidStr)
-        return this.blockstore.get(cid, callback)
+        if (has) {
+          locals.push(cid)
+        } else {
+          missing.push(cid)
+        }
+        cb()
+      })
+    }, () => {
+      if (missing.length === 0) {
+        // already finished
+        finish()
       }
 
-      addListener()
-      this.wm.wantBlocks([cid])
+      addListeners(missing)
+      this.wm.wantBlocks(missing)
+
+      this.network.findAndConnect(cids[0], CONSTANTS.maxProvidersPerRequest, (err) => {
+        if (err) {
+          this._log.error(err)
+        }
+      })
```

Comment:
What about moving this constant to the network module itself, it isn't like the API is exposed and this call can be invoked with another value.